compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs

```diff
@@ -1,21 +1,30 @@
 use std::mem::swap;
 
 use ast::HasAttrs;
+use rustc_ast::mut_visit::MutVisitor;
+use rustc_ast::visit::BoundKind;
 use rustc_ast::{
     self as ast, GenericArg, GenericBound, GenericParamKind, ItemKind, MetaItem,
     TraitBoundModifiers, VariantData,
 };
 use rustc_attr as attr;
+use rustc_data_structures::flat_map_in_place::FlatMapInPlace;
 use rustc_expand::base::{Annotatable, ExtCtxt};
 use rustc_span::symbol::{sym, Ident};
-use rustc_span::Span;
+use rustc_span::{Span, Symbol};
 use smallvec::{smallvec, SmallVec};
 use thin_vec::{thin_vec, ThinVec};
 
+type AstTy = ast::ptr::P<ast::Ty>;
```

Comment:
Please get rid of this type alias. It doesn't really make it easier to read.

Reply:
You can use `&ast::Ty` in the substitution visitor's `to_ty` field, to avoid needing to reference `P` twice. And you can just import `P` for the `visit_ty` method.

Reply:
Applied